### DISCUSSION TITLE
Automated check on the notebook formatting and PEP8

### DISCUSSION THREAD

**Discussion by ajshajib** [AstroBridge/BDLensing#127] [Announcements]
All PRs will now be automatically checked for PEP8 issues. If you see a ❌ appearing after the PR title, then the check has failed, and you need to check the error message to find out the offending issues and fix them before the PR is reviewed. If the ✅ symbol appears, your PR has passed the automated check, and only then will the PR be reviewed. This will save the PR reviewers a lot of time from checking PEP8 issues.

**The new line length limit is 88 characters, extending beyond the PEP8 limit a bit.**

Here are some instructions to quickly format your notebook before you make PRs so that it will pass the check.

1. Install `nbQA` using `python -m pip install -U nbqa`. This needs to be done only once.
2. Run `nbqa isort path_to_your_notebook.ipynb --float-to-top`.
3. Run `nbqa black path_to_your_notebook.ipynb`.
4. Run `nbqa ruff path_to_your_notebook.ipynb --fix`.

**Comment by ajshajib** [comment]
To check the error messages you have to click on "Details" as shown in the picture below, and the look for lines that look like this:
> lens_systems/DESIJ2158+0257/lens_model_DESIJ2158+0257.ipynb:cell_3:3:1: E402 Module level import not at top of file
> lens_systems/DESIJ2158+0257/lens_model_DESIJ2158+0257.ipynb:cell_3:3:8: F401 [*] `emcee` imported but unused
> lens_systems/DESIJ2158+0257/lens_model_DESIJ2158+0257.ipynb:cell_3:5:1: E402 Module level import not at top of file
> lens_systems/DESIJ2158+0257/lens_model_DESIJ2158+0257.ipynb:cell_3:6:1: E402 Module level import not at top of file
> lens_systems/DESIJ2158+0257/lens_model_DESIJ2158+0257.ipynb:cell_3:7:1: E402 Module level import not at top of file
> 

<img width="943" alt="Screenshot 2023-06-08 at 12 34 26 AM" src="https://github.com/ajshajib/BDLensing/assets/20169048/502e6c9f-de37-4810-b464-3910162b5716">

**Reply by ajshajib** [reply]
If you have too many errors that `nbqa ruff --fix` alone can't fix, you can also try using `black` and `isort` to do some additional automated formatting like this. If you have errors remaining even after then, you have to fix them by hand one by one.

```
nbqa black your_notebook.ipynb
nbqa isort your_notebook.ipynb --float-to-top
nbqa ruff your_notebook.ipynb --fix
```

@mirsazzathossain, can you kindly add the information on how to get these commands to run on Windows?

**Comment by ZareefJafar** [comment]
@ajshajib I am using [autopep8](https://github.com/hhatto/autopep8) with my JupyterLab and fixed the line length limit to 88 characters. Is it good for handling PEP8 issues? but I will still double check with nbqa.

**Reply by ajshajib** [reply]
I don't know, you have to run `nbqa ruff` to see if there are remaining issues after running `autopep8`.

**Comment by ZareefJafar** [comment]
@ajshajib Just to let you know, I was unable to find `ruff` after installing `nbqa` using  `python -m pip install -U nbqa`. Tried [`pip install ruff`](https://github.com/astral-sh/ruff#installation) now its working.
